File: Exercism/javascript/sieve/sieve.js
'use strict'

const count = function * (from) {
  while (true) {
    yield from
    from += 1
  }
}

const sieve = function * (numbers) {
  const filters = []

  while (true) {
    const { value: p } = numbers.next()

    if (filters.some(f => f(p))) continue

    yield p
    filters.push(n => n % p === 0)
  }
}

const takeWhile = function * (fn, iter) {
  let g = iter.next()
  while (!g.done) {
    if (!fn(g.value)) return
    yield g.value
    g = iter.next()
  }
}

export const primes = till =>
  [...takeWhile(n => n <= till, sieve(count(2)))]

// const log = iter => {
//   while (true) {
//     const { value, done } = iter.next()
//     if (done) break
//     console.log(value)
//   }
// }

// log(sieve(count(2)))

File: Exercism/python/sieve/sieve.py
from itertools import count


class Primes(object):
    """Iterable primes using Sieve of Eratosthenes"""
    def __init__(self, end=None):
        if not isinstance(end, (int, None)):
            raise ValueError(f'Not a valid end ({end})')
        self.numbers = count(2)
        self.end = end

    def __iter__(self):
        return self

    def __next__(self):
        p = next(self.numbers)
        if self.end is not None and p > self.end:
            raise StopIteration
        self.numbers = filter(lambda n: n % p, self.numbers)
        return p


def primes(limit: int) -> list:
    """Just to pass test"""
    return [*Primes(limit)]
